1st commit

Logan51Rus opened the pull request.

Logan51Rus committed:
1st commit

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,16 @@
+  const dispatch = useDispatch();
 
   const handleSubmit = (e: SyntheticEvent) => {
     e.preventDefault();
```
Можно лучше: более точно будет `FormEvent<HTMLFormElement>`

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,16 @@
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
+  const dispatch = useDispatch();
```
Можно лучше: Чтобы не повторять логику для подвязки инпутов можно создать хук `useForm()`, который бы позволил делать это более коротко. Пример:

```js

const [values, onChange] = useForm({email: "", password: ""})

values // {email: string, password: string}
onChange // (event) => void

```

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,16 @@
+    dispatch(loginUser({ email: email, password: password }));
   };
 
   return (
```
Можно лучше: у всех форм стоит заполнить аттрибут `autocomplete` (в реакте  `autoComplete`). Он показывает браузеру как именно следует заполнить это поле. Браузер, при наличии, может сам подставить сюда какую-то информацию. Полный список значений можно посмотреть [тут](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) (https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)

Самые полезные это: `name` или `given-name` (имя), `nickname`, `email`, `new-password`, `current-password`

kyzinatra commented on src/pages/feed/feed.tsx:
```diff
@@ -1,15 +1,31 @@
+  }, []);
+
+  console.log('orders', orders);
 
```
Можно лучше: лишние логи стоит удалить из проекта

kyzinatra commented on src/pages/feed/feed.tsx:
```diff
@@ -1,15 +1,31 @@
-  const orders: TOrder[] = [];
+  const dispatch = useDispatch();
+  const orders: TOrder[] = useSelector(selectorOrderPreview);
+
```
Можно лучше: если `useSelector` типизирован, то `orders` отдельно типизировать не нужно

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,118 @@
+        <Route
+          path='/reset-password'
+          element={
+            <ProtectedRoute onlyUnAuth>
```
Отлично! Хороший роутинг

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,118 @@
+            </Modal>
+          }
+        />
+        <Route
```
Надо исправить: модальные окна должны оставаться модальными окнами. То есть при нажатии должна быть подложка и видно содержимое страницы позади. При открытии в отдельной вкладке ссыли на модальное окно поведение может быть любым

kyzinatra commented on src/pages/login/login.tsx:
```diff
@@ -1,12 +1,16 @@
 
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
```
Можно лучше: пользователю стоит сделать вывод об ошибах в интерфейс

kyzinatra commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,118 @@
+        <Route
+          path='/ingredients/:id'
+          element={
+            <Modal title='Детали ингредиента' onClose={closeModal}>
```
Надо исправить: при перезагрузки страницы профиля выбрасывает из аккаунта

kyzinatra commented on src/services/feedSlice.ts:
```diff
@@ -0,0 +1,75 @@
+export const getOrderByNumber = createAsyncThunk(
+  'feed/getOrderByNumber',
+  async (number: number) => {
+    const res = await getOrderByNumberApi(number);
```
Отлично! Хорошая организация redux

kyzinatra commented on src/services/feedSlice.ts:
```diff
@@ -0,0 +1,75 @@
+  selectors: {
+    selectorOrderPreview: (state: TFeedState) => state.orders,
+    selectorFeedTotal: (state: TFeedState) => state.total,
+    selectorFeedTotalToday: (state: TFeedState) => state.totalToday,
```
Отлично! Круто, что вы выносите селекторы

kyzinatra commented:
Выглядит очень круто! Видно, что проделана большая работа. Но все же есть моменты на которые стоит обратить внимание.
Подробнее: 

**Что важно исправить:**

-  модальные окна должны оставаться модальными окнами. То есть при нажатии должна быть подложка и видно содержимое страницы позади. При открытии в отдельной вкладке ссыли на модальное окно поведение может быть любым
-  при перезагрузки страницы профиля выбрасывает из аккаунта


**Можно лучше:**

-  более точно будет `FormEvent<HTMLFormElement>`
-  Чтобы не повторять логику для подвязки инпутов можно создать хук `useForm()`, который бы позволил делать это более коротко. Пример:

```js

const [values, onChange] = useForm({email: "", password: ""})

values // {email: string, password: string}
onChange // (event) => void

```
-  у всех форм стоит заполнить аттрибут `autocomplete` (в реакте  `autoComplete`). Он показывает браузеру как именно следует заполнить это поле. Браузер, при наличии, может сам подставить сюда какую-то информацию. Полный список значений можно посмотреть [тут](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) (https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)

Самые полезные это: `name` или `given-name` (имя), `nickname`, `email`, `new-password`, `current-password`
-  лишние логи стоит удалить из проекта
-  если `useSelector` типизирован, то `orders` отдельно типизировать не нужно
-  пользователю стоит сделать вывод об ошибах в интерфейс


**Что понравилось:**

-  Хорошая организация redux
-  Круто, что вы выносите селекторы


**Обратите внимание, что работа принимается только после исправления всех `«Надо исправить»`.**

Logan51Rus committed:
Исправлены замечания

Logan51Rus committed:
The app was tested with cypress and jest

mikebars1995 commented on cypress/e2e/constructor/constructor.cy.tsx:
```diff
@@ -0,0 +1,84 @@
+const testUrl = 'http://localhost:4000';
```
**Можно лучше**

можно в документации посмотреть, как настроить быстро `baseUrl` для cypress https://docs.cypress.io/api/commands/visit#Visit-is-automatically-prefixed-with-baseUrl

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:


* Все юнит-тесты (35 шт) проходят успешно
* Код тестов аккуратный и понятный
* Отлично, что базовый урл вынесен в конфиг
* Отлично, что повторяющиеся селекторы вынесены в константы


но есть некоторые недочеты:

* Проект не запускается, скрин ошибки https://disk.yandex.ru/i/0R2EMlOTg-Yodg

Logan51Rus committed:
Исправлены замечания

Logan51Rus committed:
Внесены изменения

Logan51Rus committed:
Исправлены замечания

mikebars1995 reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

Logan51Rus committed:
fixed few bugs

it is possible to create order only if user is authrized; all modal windows close correctly now
